src/cargo/util/semver_ext.rs

```diff
@@ -117,9 +117,15 @@ impl OptVersionReq {
     /// and we're not sure if this part of the functionality should be implemented in semver or cargo.
     pub fn matches_prerelease(&self, version: &Version) -> bool {
         if version.is_prerelease() {
-            let mut version = version.clone();
-            version.pre = semver::Prerelease::EMPTY;
-            return self.matches(&version);
+            // Only in the case of "ordinary" version requirements with pre-release
+            // versions do we need to help the version matching. In the case of Any,
+            // Locked, or Precise, the `matches()` function is already doing the
+            // correct handling.
+            if let OptVersionReq::Req(_) = self {
+                let mut version = version.clone();
+                version.pre = semver::Prerelease::EMPTY;
+                return self.matches(&version);
+            }
```

Comment:
I don't think this is an appropriate place to be putting this fix.

Note the comment in the description

> we're not sure if this part of the functionality should be implemented in semver or cargo.

At this time, this is an abstraction that we shouldn't be breaking with cargo-specific logic

btw these two commtis seem unrelated with the rest and seem like they should be their own PR

Reply:
Hmm, but all I'm doing here is relaxing the cargo-specific logic. Without this fix, I get this bug:

```
   Upgrading pre ^0.1.0 -> ^0.2.0-beta
    Updating `dummy-registry` index
error: failed to select a version for the requirement `pre = "^0.2.0-beta"`
candidate versions found which didn't match: 0.2.0-beta
```

What do you think I should do?

Reply:
If we take `0.2.0-beta`, remove `-beta`, we get `0.2.0`.  That seems like `0.2.0-beta` should be able to match that.

Any insight into why this isn't working?

Reply:
This is a bug for this PR.

`matches_prerelease` will alaways a `OptVersionReq::Req` in my previous [PR](https://github.com/rust-lang/cargo/pull/13626),  but this PR will add `OptVersionReq::Precise` to match , see 

```

    let req = if precise.is_some() {
        OptVersionReq::Precise(new_version, new_version_req)
    } else {
        OptVersionReq::Req(new_version_req)
    };
```